daemon/ZilliqaDaemon.cpp

```diff
@@ -15,95 +15,158 @@
+      m_recovery(0),
+      m_nodeIndex(0),
+      m_syncType(0),
+      m_cseed(false) {
+  m_log = log;
+
+  if (ReadInputs(argc, argv) != SUCCESS) {
+    *m_log << "Failed to read inputs" << endl;
+    exit(EXIT_FAILURE);
+  }
 
-  logFile.open(filePath + "/" + logName);
+  *m_log << argv[0];
 
-  if (logFile.is_open()) {
-    logFile.seekg(-1, ios_base::end);
-  } else {
-    return "";
+  for (int i = 1; i < argc; ++i) {
+    *m_log << " " << argv[i];
   }
 
-  bool keepLooping = true;
+  *m_log << endl;
+  KillProcess();
+  StartNewProcess();
+  StartScripts();
+}
 
-  while (keepLooping) {
-    char ch;
-    logFile.get(ch);
+void ZilliqaDaemon::MonitorProcess() {
+  const string name = programName[0];
 
-    if ((int)logFile.tellg() <= 1) {
-      logFile.seekg(0);
-      keepLooping = false;
-    } else if (ch == ']') {
-      keepLooping = false;
-    } else {
-      logFile.seekg(-2, ios_base::cur);
+  if (m_pids[name].empty()) {
+    *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Looking for new "
+           << name << " process..." << endl;
+    vector<pid_t> tmp = ZilliqaDaemon::GetProcIdByName(name);
+
+    for (const pid_t& i : tmp) {
+      m_died[i] = false;
+      m_pids[name].push_back(i);
+      *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+             << "Started monitoring new process " << name << " with PiD: " << i
+             << endl;
     }
+
+    return;
   }
 
-  string lastLine = "";
-  getline(logFile, lastLine);
+  for (const pid_t& pid : m_pids[name]) {
+    // If sig is 0 (the null signal), error checking is performed but no signal
+    // is actually sent
+    int w = kill(pid, 0);
+
+    if (w < 0) {
+      if (errno == EPERM) {
+        *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+               << "Daemon does not have permission "
+               << "Name: " << name << " Id: " << pid << endl;
+      } else if (errno == ESRCH) {
+        *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Process died "
+               << "Name: " << name << " Id: " << pid << endl;
+        m_died[pid] = true;
+      } else {
+        *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+               << "Kill failed due to " << errno << " Name: " << name
+               << "Id: " << pid << endl;
+      }
+    }
 
-  // log<<"Line :"<<lastLine<<endl;
+    if (m_died[pid]) {
+      auto it = find(m_pids[name].begin(), m_pids[name].end(), pid);
 
-  logFile.close();
+      if (it != m_pids[name].end()) {
+        *m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Not monitoring "
+               << pid << " of " << name << endl;
+        m_pids[name].erase(it);
+      }
 
-  return lastLine;
+      StartNewProcess();
+      m_died.erase(pid);
+    }
+  }
 }
 
-void setIsDs(const pid_t& pid, ofstream& log) {
-  string s = ReadLastLine(Path[pid], log);
+string ZilliqaDaemon::CurrentTimeStamp() {
+  auto tm = chrono::system_clock::to_time_t(chrono::system_clock::now());
+  char* t = ctime(&tm);
```

Comment:
Perhaps a better choice is `std::strftime` so the manual string operation can be avoided.

See [`std::strftime`](https://en.cppreference.com/w/cpp/chrono/c/strftime) and [`std::ctime`](https://en.cppreference.com/w/cpp/chrono/c/ctime)
> POSIX marks this function obsolete and recommends std::strftime instead.